os/drivers/video/video_framebuff.c

```diff
@@ -0,0 +1,264 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include "video_framebuff.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+
+#include <tinyara/irq.h>
+#include <tinyara/kmalloc.h>
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static void init_buf_chain(video_framebuff_t *fbuf)
+{
+	int i;
+	vbuf_container_t *tmp;
+
+	fbuf->vbuf_empty = fbuf->vbuf_alloced;
+	fbuf->vbuf_next_dma = NULL;
+	fbuf->vbuf_dma = NULL;
+	fbuf->vbuf_top = NULL;
+	fbuf->vbuf_tail = NULL;
+
+	tmp = fbuf->vbuf_alloced;
+	for (i = 0; i < fbuf->container_size - 1; i++) {
+		tmp->next = &tmp[1];
+		tmp++;
+	}
+}
+
+static void cleanup_container(video_framebuff_t *fbuf)
+{
+	if (fbuf->vbuf_alloced) {
+		memset(fbuf->vbuf_alloced, 0, sizeof(vbuf_container_t) * fbuf->container_size);
+		init_buf_chain(fbuf);
+	}
+}
+
+static inline int is_last_one(video_framebuff_t *fbuf)
+{
+	return fbuf->vbuf_top == fbuf->vbuf_tail ? 1 : 0;
+}
+
+static inline vbuf_container_t *dequeue_vbuf_unsafe(video_framebuff_t *fbuf)
+{
+	vbuf_container_t *ret = fbuf->vbuf_top;
+	if (is_last_one(fbuf)) {
+		fbuf->vbuf_top = NULL;
+		fbuf->vbuf_tail = NULL;
+		fbuf->vbuf_next_dma = NULL;
+	} else {
+		if (fbuf->mode == V4L2_BUF_MODE_RING) {
+			fbuf->vbuf_tail->next = fbuf->vbuf_top->next;
+		}
+		fbuf->vbuf_top = fbuf->vbuf_top->next;
+	}
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+void video_framebuff_init(video_framebuff_t *fbuf)
+{
+	fbuf->mode = V4L2_BUF_MODE_RING;
+	fbuf->vbuf_empty = NULL;
+	fbuf->vbuf_top = NULL;
+	fbuf->vbuf_tail = NULL;
+	fbuf->vbuf_next_dma = NULL;
+
+	sem_init(&fbuf->lock_empty, 0, 1);
+}
+
+void video_framebuff_uninit(video_framebuff_t *fbuf)
+{
+	video_framebuff_realloc_container(fbuf, 0);
+	sem_destroy(&fbuf->lock_empty);
+}
+
+int video_framebuff_realloc_container(video_framebuff_t *fbuf, int sz)
+{
+	if (sz > V4L2_REQBUFS_COUNT_MAX) {
+		return -EINVAL;
+	}
+
+	if (fbuf->vbuf_alloced == NULL || fbuf->container_size != sz) {
```

Comment:
If container_size is same as sz, then just cleanup?
```
/* First free already allocated buf if size is different */
if ((fbuf->vbuf_allocated != NULL)  {
       /* Now, check realloc condition */
       if ((fbuf->container_size != sz) && (sz > 0)) {
            kmm_free(fbuf->vbuf_alloced);
       } else { /* already same size, then just clean up or return */
           goto cleanup;
      }
}

fbuf->vbuf_allocated = (vbuf_container_t *)kmm_malloc(sizeof(vbuf_container_t) * sz);
..
cleanup :
          cleanup_container(fbuf);
          return OK;
```

Reply:
Corrected